can-observe.js

```diff
@@ -1,35 +1,211 @@
-
 var cid = require("can-cid");
-var canEvent = require("can-event");
 var canBatch = require("can-event/batch/batch");
 var Observation = require("can-observation");
+var canSymbol = require("can-symbol");
+var canReflect = require("can-reflect");
+var namespace = require("can-namespace");
+var canEvent = require("can-event");
 
-var has = Object.prototype.hasOwnProperty;
+var observableSymbol = canSymbol.for("can.observeData");
+var hasOwn = Object.prototype.hasOwnProperty;
+
+function isIntegerIndex(prop) {
+	return prop && // empty string typecasts to 0
+		(+prop === +prop) && // NaN check for strings that don't represent numbers
+		(+prop % 1 === 0);  // floats should be treated as strings
+}
+
+var proxyOnly = Object.create(null);
+canReflect.assignSymbols(proxyOnly, {
+	"can.onKeyValue": function(key, handler) {
+		var handlers = this[observableSymbol].handlers;
+		var keyHandlers = handlers[key];
+		if (!keyHandlers) {
+			keyHandlers = handlers[key] = [];
+		}
+		// TODO: Binding directly on getters (currently does not work)
+		//   need to add an observation instead of just firing handlers.
+		keyHandlers.push(handler);
+	},
+	"can.offKeyValue": function(key, handler) {
+		var handlers = this[observableSymbol].handlers;
+		var keyHandlers = handlers[key];
+		if(keyHandlers) {
+			var index = keyHandlers.indexOf(handler);
+			if(index >= 0) {
+				keyHandlers.splice(index, 1);
+			}
+		}
+	}
+});
+canReflect.assign(proxyOnly, canEvent);
+
+var arrayMethodInterceptors = Object.create(null);
+var mutateMethods = {
+	"push": {
+		add: function(arr, args, retVal) {
+			return [[args, arr.length - args.length]];
+		}
+	},
+	"pop": {
+		remove: function(arr, args, retVal) {
+			return [[[retVal], arr.length]];
+		}
+	},
+	"shift": {
+		remove: function(arr, args, retVal) {
+			return [[[retVal], 0]];
+		}
+	},
+	"unshift": {
+		add: function(arr, args, retVal) {
+			return [[args, 0]];
+		}
+	},
+	"splice": {
+		remove: function(arr, args, retVal) {
+			return [[retVal, args[0]]];
+		},
+		add: function(arr, args, retVal) {
+			return [[args.slice(2), args[0]]];
+		}
+	},
+	"sort": {
+		move: function(arr, args, retVal, old) {
+			return arr.map(function(element, index) {
+				if(old[index] !== element) {
+					return [element, index, old.indexOf(element)];
+				}
+			}).filter(function(el) {
```

Comment:
is this `n^2`?

Reply:
what is this doing? Firing move events?  We should probably not doing that unless someone installs a special sort plugin.  I'd instead do what `can-define/list` and `can-list` do ... events that add/remove the contents of the list.

Reply:
It looks like the way I had it implemented wasn't doing it right anyway (things just flipped back and forth and nothing actually changed). There's nothing strictly wrong with doing move events here (because can-view-live can use them), but that would require using our existing array diffing.  Anyway, it's a task for another time.  I now have it doing complete removal and addition.